Make it possible to not write standard annotation output

akotlar opened the pull request:
* Add noOut option which prevents writing the annotation to stdout/file. If dosageOut is not provided (no dosage requested), noOut can't be used 
* Fixes storing nil dosage for missing genotype in multi allelic case
* Pass arrow ipc.Option more idiomatically
* Rename NewArrowWriter to NewArrowIPCFileWriter for clarity

akotlar commented on main.go:
```diff
@@ -555,99 +564,116 @@ func processLines(header []string, numChars int, config *Config, queue chan [][]
-				} else {
-					output.WriteString(parse.GetTrTv(string(refs[i]), alts[i]))
-				}
+				if !config.noOut {
```
from here to the end of the function has not been changed, just indented due to the new condition (!config.noOut)

wingolab reviewed (commented):
@akotlar  - 

1) please consider cleaning up your test artifacts - e.g., defer os.Remove() or use a tempfile / temp directory when you're writing/reading files

2) I'd write a func called `NewArrowIPCFWriter` that takes an io.writer and use it for testing. In the package, `NewArrowIPCFileWriter` (or, my suggested rename is `NewArrowIPCWriterToFile`) would open the file, call `NewArrowIPCFWriter` and close the writer For testing, you could directly test the data in buffer of whatever underlies the  io.Writer rather than needing to hit the disk for tests, and your testing would be easier to follow since you could pull the logic from `readAndVerifyArrowFile` into the test itself (`readArrowRows` could use a io.reader). Right now, `readAndVerifyArrowFile` deviates from the idiomatic approach for failing tests. If you wanted to be more idiomatic but keep `readAndVerifyArrowFile`, you'd pass the pointer to testing.T.

akotlar commented:
> 2\. I'd write a func called `NewArrowIPCFWriter` that takes an io.writer and use it for testing. In the package, `NewArrowIPCFileWriter` (or, my

I addressed your comments. However, I cannot directly follow the NewArrowIPCFWriter approach for testing. The arrow library NewFileWriter interface is `func NewFileWriter(w io.WriteSeeker, opts ...Option) (*FileWriter, error)`. io.WriteSeeker has no in-memory implementation: https://stackoverflow.com/questions/45836767/using-an-io-writeseeker-without-a-file-in-go. I chose to follow the approach that the Arrow Go library maintainers took, which directly tests with file creation:

- https://github.com/apache/arrow/blob/main/go/arrow/ipc/file_test.go#L37
- https://github.com/apache/arrow/blob/main/go/arrow/internal/arrdata/ioutil.go#L145

wingolab commented on arrow/arrow_test.go:
```diff
@@ -260,14 +272,22 @@ func TestArrowWriterHandlesNullValues(t *testing.T) {
+	os.Remove(filePath)
 }
 
 func runConcurrentTest(compress bool) {
```
Where is `compress` used? I deleted it and re-ran tests without changing anything.

akotlar replied:
fixed

wingolab reviewed (commented):
Your explanation about the underlying library makes how you setup things easier to understand. Since there's no in-memory io.WriteSeeker in the std library it makes sense not to implement our own. I had a question about the compression tests.

```diff
:100644 100644 396dce9 0000000 M	arrow/arrow_test.go
:100644 100644 4541310 0000000 M	main_test.go

diff --git a/arrow/arrow_test.go b/arrow/arrow_test.go
index 396dce9..79ef5eb 100644
--- a/arrow/arrow_test.go
+++ b/arrow/arrow_test.go
@@ -138,11 +138,11 @@ func readArrowRows(filePath string) ([][]any, error) {
 	return readRows, nil
 }
 
-func readAndVerifyArrowFile(filePath string, expectedRows [][]any, sort bool) {
+func checkArrowFile(t *testing.T, filePath string, expectedRows [][]any, sort bool) {
 	readRows, err := readArrowRows(filePath)
 
 	if err != nil {
-		log.Fatal(err)
+		t.Fatal(err)
 	}
 
 	if sort {
@@ -152,7 +152,7 @@ func readAndVerifyArrowFile(filePath string, expectedRows [][]any, sort bool) {
 
 	for i, row := range readRows {
 		if !reflect.DeepEqual(row, expectedRows[i]) {
-			log.Fatalf("Mismatch at row %d: got %v, want %v", i, row, expectedRows[i])
+			t.Fatalf("Mismatch at row %d: got %v, want %v", i, row, expectedRows[i])
 		}
 	}
 }
@@ -197,7 +197,7 @@ func TestArrowWriteRead(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	readAndVerifyArrowFile(filePath, rows, false)
+	checkArrowFile(t, filePath, rows, false)
 
 	os.Remove(filePath)
 }
@@ -271,12 +271,12 @@ func TestArrowWriterHandlesNullValues(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	readAndVerifyArrowFile(filePath, rows, false)
+	checkArrowFile(t, filePath, rows, false)
 
 	os.Remove(filePath)
 }
 
-func runConcurrentTest(compress bool) {
+func conncurentTestMaybeCompress(t *testing.T, compress bool) {
 	filePath := filepath.Join(os.TempDir(), "concurrent_output.feather")
 	file, err := os.Create(filePath)
 	if err != nil {
@@ -334,17 +334,17 @@ func runConcurrentTest(compress bool) {
 		log.Fatal(err)
 	}
 
-	readAndVerifyArrowFile(filePath, rows, true)
+	checkArrowFile(t, filePath, rows, true)
 
 	os.Remove(filePath)
 }
 
 func TestArrowWriterConcurrency(t *testing.T) {
-	runConcurrentTest(false)
+	conncurentTestMaybeCompress(t, false)
 }
 
 func TestArrowWriterConcurrencyWithCompression(t *testing.T) {
-	runConcurrentTest(true)
+	conncurentTestMaybeCompress(t, true)
 }
 
 func TestNewArrowIPCFileWriterWithZstdOption(t *testing.T) {
diff --git a/main_test.go b/main_test.go
index 4541310..444fb0c 100644
--- a/main_test.go
+++ b/main_test.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"log"
 	"os"
+	"path"
 	"path/filepath"
 	"strconv"
 	"strings"
@@ -168,7 +169,8 @@ func TestHeader(t *testing.T) {
 }
 
 func TestWriteSampleListIfWanted(t *testing.T) {
-	config := Config{sampleListPath: "./test.sample_list"}
+	filepath := path.Join(os.TempDir(), "./test.sample_list")
+	config := Config{sampleListPath: filepath}
 
 	versionLine := "##fileformat=VCFv4.x"
 	header := strings.Join([]string{"#CHROM", "POS", "ID", "REF", "ALT", "QUAL",
@@ -188,7 +190,7 @@ func TestWriteSampleListIfWanted(t *testing.T) {
 
 	readVcf(&config, reader, w)
 
-	inFh, err := os.Open("./test.sample_list")
+	inFh, err := os.Open(filepath)
 	if err != nil {
 		log.Fatal(err)
 	}
@@ -225,7 +227,8 @@ func TestWriteSampleListIfWanted(t *testing.T) {
 }
 
 func TestWriteSampleListWhenNoSamples(t *testing.T) {
-	config := Config{sampleListPath: "./test.sample_list_2"}
+	filepath := path.Join(os.TempDir(), "test.sample_list_2")
+	config := Config{sampleListPath: filepath}
 
 	versionLine := "##fileformat=VCFv4.x"
 	header := strings.Join([]string{"#CHROM", "POS", "ID", "REF", "ALT", "QUAL",
@@ -245,7 +248,7 @@ func TestWriteSampleListWhenNoSamples(t *testing.T) {
 
 	readVcf(&config, reader, w)
 
-	inFh, err := os.Open("./test.sample_list_2")
+	inFh, err := os.Open(filepath)
 	if err != nil {
 		log.Fatal(err)
 	}
```

wingolab reviewed (approved):
Since the intention of this library is to specifically support bystro, the changes make sense. Updating the logic for dealing with VCF parsing (without deeply tying it to using arrow) would make it more useful for situations that are not in the Bystro ecosystem and, also, consider adding a github action to test and build the binary. Both could be added to the roadmap or plans for the project.
